metis-transformation/metis-transformation-service/src/main/java/eu/europeana/metis/transformation/service/CacheWithExpirationTime.java

```diff
@@ -0,0 +1,219 @@
+
+    // If we have the item and it is still valid, we return it.
+    lock.readLock().lock();
+    try {
+      if (cache.containsKey(key)) {
+        final CacheItemWithExpirationTime<V> cachedItem = cache.get(key);
+        return supplier -> cachedItem.getValue(expirationTime, supplier, lenientWithReloads);
+      }
+    } finally {
+      lock.readLock().unlock();
+    }
+
+    // We are still here: so either we don't have the item or it is no longer valid.
+    lock.writeLock().lock();
+    try {
+
+      // Recheck the state: maybe the item was added while we waited for the write lock.
+      if (cache.containsKey(key)) {
+        final CacheItemWithExpirationTime<V> cachedItem = cache.get(key);
+        return supplier -> cachedItem.getValue(expirationTime, supplier, lenientWithReloads);
+      }
+
+      // So we need to add the cache item.
+      final CacheItemWithExpirationTime<V> result = new CacheItemWithExpirationTime<>();
+      cache.put(key, result);
+      return supplier -> result.getValue(expirationTime, supplier, lenientWithReloads);
+
+    } finally {
+      lock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Obtain the value for the given key. If the value is not present, or has expired, the supplier
+   * is used to get the latest version of the value (and store it in cache).
+   * 
+   * @param key The key to look up.
+   * @param valueSupplier A supplier that can be used to obtain the latest version of the value.
+   * @return The value.
+   * @throws CacheValueSupplierException In case a problem occurred while obtaining the latest
+   *         version of the value.
+   */
+  public V getFromCache(K key, CacheValueSupplier<V> valueSupplier)
+      throws CacheValueSupplierException {
+    return getValueResolver(key).apply(valueSupplier);
+  }
+
+  /**
+   * <p>
+   * This method cleans the cache of any items that have not been accessed for a given amount of
+   * time (i.e. in the time span given by the parameter). This method could be called by a scheduled
+   * cleanup.
+   * </p>
+   * <p>
+   * NOTE: It is theoretically possible that a cached item is being removed that is also currently
+   * in the process of being requested. This is considered to be unlikely and warranting neither the
+   * extra locking (i.e. performance hit) nor the more complex code that would need to be in place
+   * to prevent this. Even if this does occur, the behavior of this cache remains unchanged. The
+   * only consequence is that the requested value may need to be reloaded again when it is next
+   * requested.
+   * </p>
+   * 
+   * @param since The interval length of the period we want to check (which ends now). A negative
+   *        duration cleans everything.
+   */
+  public void removeItemsNotAccessedSince(Duration since) {
+    lock.writeLock().lock();
+    try {
+      final Iterator<Entry<K, CacheItemWithExpirationTime<V>>> iterator =
+          cache.entrySet().iterator();
+      while (iterator.hasNext()) {
```

Comment:
Have you heard of removeIf method?
this block could be simply for example:
cache.entrySet().removeIf(entry -> !entry.getValue().valueWasAccessedRecently(since));